webapp/tests/Controller/ControllerRolesTest.php

```diff
@@ -0,0 +1,221 @@
+<?php declare(strict_types=1);
+
+namespace App\Tests\Controller;
+
+use App\Tests\BaseTest;
+
+class ControllerRolesTest extends BaseTest
+{
+    protected static $roles = [];
+
+    /**
+     * //See: https://www.oreilly.com/library/view/php-cookbook/1565926811/ch04s25.html
+     * Get all combinations of roles with at minimal the starting roles
+     * @return array $results
+     * @var string[] $possible_roles
+     * @var string[] $start_roles
+     */
+    protected function roleCombinations(array $start_roles, array $possible_roles)
+    {
+        // initialize by adding the empty set
+        $results = array($start_roles);
+
+        foreach ($possible_roles as $element) {
+            foreach ($results as $combination) {
+                array_push($results, array_merge(array($element), $combination));
+            }
+        }
+        return $results;
+    }
+
+    /*
+     * Some URLs are not setup in the testing framework or have a function for the
+     * user UX/login process, those are skipped.
+     * @var string $url
+     * @return boolean $includedInTest
+     */
+    public function urlExcluded(string $url)
+    {
+        // Documentation is not setup in the UnitTesting framework
+        if (substr($url, 0, 4) == '/doc') {
+            return true;
+        }
+        // API is not functional in Testing framework
+        if (substr($url, 0, 4) == '/api') {
+            return true;
+        }
+        // The change-contest handles a different action
+        if (substr($url, 0, 21) == '/jury/change-contest/') {
+            return true;
+        }
+        // Remove links to local page, external or application links
+        if ($url[0] == '#' || $url[0] == 'h' || $url == '/logout') {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Crawl the webpage assume this is allowed and return all other links on the page
+     * @return string[] $urlsToCheck
+     * @var string $url
+     * @var int $statusCode
+     */
+    public function crawlPage(string $url, int $statusCode)
+    {
+        // We exclude some possible urls to not break the MockData
```

Comment:
Why are there two plaes in this file that list URL parts to exclude (also urlExclude()) with somewhat the same content?

Reply:
Yeah I'd say we should try to merge these two cases? I think it doesn't hurt to exclude the same URL's in both cases?

Reply:
By merging you would remove the failing parts of one test also from the other test where this might not fail right? Its definitely doable and I suspect that our tests are run on the same Mockdata? So one test changing state will have effect on other tests?